Iter1 new

jbakhtin opened the pull request.

Commit:
Init commit

jbakhtin committed:
Update README.md

jbakhtin committed:
Update README.md

jbakhtin committed:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
create docs/

jbakhtin committed:
create diagram without name.drawio.png

Commit:
Changes:
- rename image name from cyrillic to latin
- add db structure image to README.md

Commit:
update README.md

Commit:
delete .github/workflows/gophermart.yml file

Commit:
Changes:
- added handlers

Commit:
update README.md

jbakhtin committed:
update db-structure.drawio.png

Commit:
update README.md

jbakhtin committed:
Update db-structure.drawio.png

Commit:
Features:
- added migrations for for users table
- added handlers
- added server
- added connection to db

Commit:
update README.md

Commit:
Update README.md

Commit:
change GetById to GetByID

Commit:
change Id to ID

Commit:
set status code into handler mark Update()

Commit:
Changes:
- changed structure of progect
- added diy auth with redis

Commit:
Features:
- added jwt auth
- register
- login
- check auth middleware

Commit:
Changed:
- auth with jwt-go

Commit:
Merge branch 'iter-1-auth' into iter1

Commit:
Changes:
- added error handlers

Commit:
Refactoring:
- changes related with style

jbakhtin committed:
Update db-structure.drawio.png

Commit:
Changes:
- goimports

Commit:
Features:
- use config to auth

Commit:
Features:
- added test for auth handler (Registration and Autentification)

Commit:
Changes:
- added tests for auth handler
- changed apperror handler
- changed design of routes

Commit:
Changes:
- closed response body in test file

Commit:
Changes:
- added req.Body.Close() for tests

Commit:
Changes:
- changde tests

Commit:
Updated README.md

Commit:
Features:
- added email worker pool
- bility to send messages

Commit:
Updated README.md

Commit:
Changes:
- move repo, service and handler creation to composite

tmvrus commented on cmd/routecoverage/main.go:
```diff
@@ -0,0 +1,67 @@
+func main() {
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		fmt.Println(err)
```
fmt не рекомендуется использовать для логирования, это материал 3го спринта

jbakhtin replied:
Поправил, логирую с помощью стандартного логера в случае, если zap вернул ошибку

tmvrus commented on cmd/routecoverage/main.go:
```diff
@@ -0,0 +1,67 @@
+
+	cfg := config.GetConfig()
+	if err != nil {
+		logger.Error(err.Error())
```
тут какая-то странная ситуация: выглидят так что config.GetConfig() должен возвращать ошибку, но не возвращает. Какая тогда ошибка проверятся сразу после?

jbakhtin replied:
Поправил, добавил ошибку возвращаемую вторым значением

tmvrus commented on cmd/routecoverage/main.go:
```diff
@@ -0,0 +1,67 @@
+	go func() {
+		if err = myServer.Start(); err != nil {
+			logger.Info(err.Error())
+			return
```
этот return лишен смысла, если его убрать, ничего не изменится

jbakhtin replied:
Поправил

tmvrus commented on cmd/routecoverage/main.go:
```diff
@@ -0,0 +1,67 @@
+		}
+	}()
+
+	ctxOS, _ := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGINT, syscall.SIGQUIT)
```
рекомендую передать тут внешний контекст, поскольку он будет учитываться при обработке сигналов
рекомендую ознакомиться
https://pkg.go.dev/os/signal#NotifyContext

jbakhtin replied:
Переписал завершение программы. Реализовал пакет Closer который позволяет передать в него все функции завершения

tmvrus commented on cmd/routecoverage/main.go:
```diff
@@ -0,0 +1,67 @@
+	<-ctxOS.Done()
+	err = myServer.Shutdown(ctxServer)
+	if err != nil {
+		logger.Info(err.Error())
```
довольно странно логировать ошибку с уровнем Info

jbakhtin replied:
Поправил

tmvrus commented on cmd/routecoverage/main.go:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	cancel()
+	time.Sleep(2 * time.Second)
```
смысл этого кода понятен только автору, для меня это выглядит как костыль под неудачный дизайн остановки приложения

jbakhtin replied:
Переписал завершение программы. Перенес ожидание окончания функций завершения в пакет Сloser

tmvrus commented on internal/application/app.go:
```diff
@@ -0,0 +1,69 @@
+
+	r.NotFound(notFound())
+
+	authComposite, _ := api.NewAuthComposite(cfg)
```
не ясно почему тут игнорируется возвращаемая ошибка

jbakhtin replied:
Действительно, упустил. Добавил обработку ошибки

tmvrus commented on internal/application/app.go:
```diff
@@ -0,0 +1,69 @@
+
+		r.Get("/test", func(writer http.ResponseWriter, request *http.Request) {
+			fmt.Println("test")
+		})
```
если нужен тестовый хендлер который напишет в браузер test, то лучше использовать writer.Write
если нужно логировать, то используй логгер

tmvrus commented on internal/application/app.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+func (s *Server) Start() error {
+	err := s.ListenAndServe()
```
ListenAndServe всегда возвращает ошибку, рекомендую ознакомится с документацией и обрабатывать это корректно

jbakhtin replied:
Поправил

tmvrus commented on internal/application/apperror/error.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+var (
+	SystemErrorCode       = "000"
+	NotFoundCode          = "001"
+	UserAlreadyExistsCode = "002"
+	BadRequestParamsCode  = "003"
```
вот эти коды выглядят довольно странно, рекомендую еще раз ознакомиться с разделом errors курса

jbakhtin replied:
Это уникальные идентификаторы ошибок, по крайней мере я постарался это реализовать (Или хотя бы заложить такую возможность, поэкспериментировать с ней). Они предназначены для возврата клиенту, используещиму наше API. Например UserAlreadyExistsCode или BadRequestParamsCode будут возвращены со статусом http BadRequest 400, но благодаря идентификатору ошибки они смогут найти более подробную информацию о ней в документации к API.

tmvrus commented on internal/domain/repositories/user.go:
```diff
@@ -0,0 +1,20 @@
+	CreateUser(user UserRegistration) (*models.User, error)
+	UpdateUser() (bool, error)
+	DeleteUser() (bool, error)
+}
```
в методы базы данных принято пробрасывать внешний контекст из запроса

tmvrus commented on internal/domain/services/auth.go:
```diff
@@ -0,0 +1,140 @@
+func (ulr *UserLoginResponse) Marshal() []byte {
+	marshal, err := json.Marshal(ulr)
+	if err != nil {
+		return nil
```
ошибку можно хотя бы логировать

jbakhtin replied:
да, странное решение, переписал - прокидываю ошибку ниже и логирую

tmvrus commented on internal/domain/services/auth.go:
```diff
@@ -0,0 +1,140 @@
+	// TODO: find user
+	user, err := us.repo.GetUserByLogin(userLogin.Login)
+	if err != nil {
+		if strings.Contains(err.Error(), "sql: no rows in result set") {
```
errors.Is + sql.ErrNoRows

jbakhtin replied:
Поправил. Почему константы есть не для всех ошибок? Например я не могу найти константу для ошибки duplicate keys

tmvrus commented on internal/domain/services/auth.go:
```diff
@@ -0,0 +1,140 @@
+		return nil, err
+	}
+
+	h := hmac.New(sha256.New, []byte(us.config.AppKey))
+	h.Write([]byte(fmt.Sprintf("%s:%s", userLogin.Login, userLogin.Password)))
+	hashedPassword := h.Sum(nil)
```
это довольно чувствительный код, лучше выделать его в отдельную ф-ю

tmvrus commented on internal/infrastructure/api/handlers/area.go:
```diff
@@ -0,0 +1,43 @@
+package handlers
+
+//func (h *Handler) CreateArea() http.HandlerFunc {
```
в рабочем проекте не должно быть кода который закомментирован

tmvrus commented on internal/infrastructure/api/handlers/auth.go:
```diff
@@ -0,0 +1,91 @@
+
+		registerResponse, err := h.service.RegisterUser(request)
+		if err != nil {
+			if strings.Contains(err.Error(), "duplicate key value violates") {
```
errors.Is

jbakhtin replied:
Переделал, пред сосзданием пользваотеля делаю дополнительный запрос к базе данных, на проверку налисчия  пользователя.

tmvrus commented on internal/infrastructure/api/handlers/mark.go:
```diff
@@ -0,0 +1,43 @@
+package handlers
+
+//func (h *Handler) CreateMark() http.HandlerFunc {
```
в рабочем проекте не должно быть кода который закомментирован

tmvrus commented on internal/infrastructure/api/handlers/route.go:
```diff
@@ -0,0 +1,43 @@
+package handlers
+
+//func (h *Handler) CreateRoute() http.HandlerFunc {
```
в рабочем проекте не должно быть кода который закомментирован

tmvrus commented on internal/infrastructure/api/middleware/validateLoginParams.go:
```diff
@@ -0,0 +1,42 @@
+
+		errsList := map[string]string{}
+		request := services.UserLoginRequest{}
+		_ = json.Unmarshal(bodyBytes, &request)
```
игнорирование ошибки

jbakhtin replied:
поправил

tmvrus commented on internal/infrastructure/api/middleware/validateLoginParams.go:
```diff
@@ -0,0 +1,42 @@
+		req := r.Clone(r.Context())
+
+		bodyBytes, _ := io.ReadAll(req.Body)
+		req.Body.Close() //  must close
```
игнорирование ошибки

jbakhtin replied:
Поправил

tmvrus commented on internal/infrastructure/api/middleware/validateRegistrationParams.go:
```diff
@@ -0,0 +1,62 @@
+
+func ValidateRegistrationParams(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) error {
+		req := r.Clone(r.Context())
```
какой в этом смысл?

jbakhtin replied:
Так как тело запроса можно прочитать только один раз, следовательно мне необходимо его скопировать, что бы произвести над ним некоторые действия в middleware, а потом исходный запрос передать для дальнейшей обработки в handler

tmvrus commented on internal/infrastructure/database/postgres/repository/user.go:
```diff
@@ -0,0 +1,87 @@
+
+func (ur *UserRepository) CreateUser(user repositories.UserRegistration) (*models.User, error) {
+	var stored models.User
+	err := ur.QueryRow(query.CreateUser, &user.Name, &user.Lastname, &user.Login, &user.Email, &user.Password).
```
как обрабатывается случай когда пользователь уже есть?

jbakhtin replied:
На уровне сервисного слоя я делаю обращение к базе данный и проверяю наличие пользователя с помощью метода getUserByLogin()

tmvrus commented on internal/infrastructure/database/postgres/repository/user.go:
```diff
@@ -0,0 +1,87 @@
+
+func (ur *UserRepository) GetUserByID(id int) (*models.User, error) {
+	var user models.User
+	err := ur.QueryRow(query.GetUserByID, id).
```
современный подход - использовать QueryRowContext вместо QueryRow

jbakhtin replied:
Поправил

tmvrus commented on internal/infrastructure/database/postgres/repository/user.go:
```diff
@@ -0,0 +1,87 @@
+	return &user, nil
+}
+
+func (ur *UserRepository) GetUsers() ([]models.User, error) {
+	return nil, nil
+}
+
+func (ur *UserRepository) UpdateUser() (bool, error) {
+	return false, nil
+}
+
+func (ur *UserRepository) DeleteUser() (bool, error) {
+	return false, nil
+}
```
тут печальная пустота

tmvrus commented on internal/interfaces/services/auth.go:
```diff
@@ -0,0 +1,10 @@
+)
+
+type AuthService interface {
+	RegisterUser(request services.UserRegistrationRequest) (*services.UserRegistrationResponse, error)
+	LoginUser(request services.UserLoginRequest) (*services.UserLoginResponse, error)
```
не хватает контекста

tmvrus commented on internal/utils/mailer/mailer.go:
```diff
@@ -0,0 +1,86 @@
+	Body string
+}
+
+func NewMail(to string, subject string, body string) *Mail {
```
go позволяет указать тип один раз
```
func NewMail(to, subject, body string) *Mail
```

jbakhtin replied:
Поправил

tmvrus commented on internal/utils/mailer/mailer.go:
```diff
@@ -0,0 +1,86 @@
+
+func GetMailsQueue() chan Mail {
+	once.Do(func() {
+		mails = make(chan Mail, 5)
```
смысл цифры 5 понятен только автору кода, если бы тут была именованная константа - было бы читабельней

jbakhtin replied:
перенес в кофниг

tmvrus commented on internal/utils/mailer/mailer.go:
```diff
@@ -0,0 +1,86 @@
+			go func() {
+				err := m.Send(&mail)
+				if err != nil {
+					fmt.Println(err)
```
log

tmvrus commented on internal/utils/mailer/mailer.go:
```diff
@@ -0,0 +1,86 @@
+
+	err := smtp.SendMail(smtpAddr, auth, m.config.Mail.FromAddress, []string{mail.To}, msg)
+	if err != nil {
+		fmt.Println(err)
```
log

tmvrus commented on internal/utils/mailer/mailer.go:
```diff
@@ -0,0 +1,86 @@
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
```
если в канале mails останутся письма, то они никуда не отправятся, код к этому не готов

jbakhtin replied:
да, нужно будет доработать и сохранить письма перед завершением прогрммы

tmvrus commented on internal/utils/ratelimiter/ratelimiter.go:
```diff
@@ -0,0 +1,57 @@
+package ratelimiter
```
весь этот код можно заменить одним каналом с нужной емкостью:
```go
limiter := make(chan struct{}, limit)
limiter <- struct{} // захватываем слот
// тут выполняется работа
<- limiter // освобождаем слот

```

tmvrus reviewed (changes requested):
- у тебя индивидуальный проект, в нем лежат бинарный файлы из общего
- рекомендую еще раз ознакомиться с темой контекстов
- не следует игнорировать ошибки, их следует либо логировать либо пробрасывать в вызывающий код
- код которые не используется - засоряет проект и усложняет его чтение

Commit:
Updated app logger
- added ability save logs to file
- added autortate
- added split to prod and dev

Commit:
Features:
Added a CLouser package that allows to gracefully shut down processes

Commit:
Changes:
- changed handle errors ErrNoRows and ErrDuplicteKey

Commit:
Refactoring

Commit:
Changes:
- fixed errors from code linter

Commit:
Fixes:
- fixed auth_test.go

Commit:
Changes:
- added context
- changed test route

Commit:
Changes:
- added routes table
- added routes creation endpoint

Commit:
Updated README.md

Commit:
Changes:
- added ability to store route
- added ability to  route

Commit:
Changes:
- added ability to update Route

Commit:
Changes:
- fixeds tags of struct
- changed Id to ID

Commit:
Update README.md

Commit:
Refactoring:
- added forgoten error handlers

Commit:
Changes:
- added check authoriztion when make requests
- changed sql queriew related with user_id
- added user_id into context

Commit:
Features:
- delete routes by Id

Commit:
Changes:
- added context key

Commit:
Refactoring:
- moved context type to derictory

Commit:
Features:
- added ability to get routes list

Commit:
Changes:
- added rows.Err() check

Commit:
Updated README.md

jbakhtin committed:
Updated db-structure.drawio.png

jbakhtin commented:
Merge iteration 1 to Main
